### PR TITLE
Allowed players to eat in creative mode and difficulty peaceful

### PR DESCRIPTION
**Pull request by NTT1906**
## Introduction
Allowed players to eat in creative mode and difficulty peaceful.

- Added **`Player::canEatWithoutHunger()`.**

## Backwards compatibility
Maybe?

## Tests

https://github.com/pmmp/PocketMine-MP/assets/54394881/961b72ef-62e2-4e46-a470-844cf21482e9

### CODE REVIEW

File: src/entity/Human.php [hunk 1]
```diff
@@ -190,11 +190,10 @@ public function getHungerManager() : HungerManager{
 		return $this->hungerManager;
 	}
 
-	public function consumeObject(Consumable $consumable) : bool{
-		if($consumable instanceof FoodSource && $consumable->requiresHunger() && !$this->hungerManager->isHungry()){
+	public function consumeObject(Consumable $consumable, bool $forceEating = false) : bool{
+		if (($consumable instanceof FoodSource) && !$forceEating && $consumable->requiresHunger() && !$this->getHungerManager()->isHungry()){
```
Comment (ShockedPlot7560) [inline src/entity/Human.php]: Perhaps you should put it on several lines to make the condition more legible.

File: src/player/Player.php [hunk 2]
```diff
@@ -673,6 +674,13 @@ public function setUsingItem(bool $value) : void{
 		$this->networkPropertiesDirty = true;
 	}
 
+	/**
+	 * Allow eating
+	 */
+	public function forceAllowEating() : bool{
```
Comment (ShockedPlot7560) [inline src/player/Player.php]: ``allowForceEating`` seems better

File: src/player/Player.php [hunk 3]
```diff
@@ -2000,6 +2008,10 @@ public function emote(string $emoteId) : void{
 		}
 	}
 
+	public function consumeObject(Consumable $consumable, bool $forceEating = false) : bool{
+		return parent::consumeObject($consumable, $this->forceAllowEating() || $forceEating);
```
Comment (ShockedPlot7560) [inline src/player/Player.php]: Prefer conditions that consume less first, i.e. check the variable before the function.

File: src/entity/Human.php [hunk 4]
```diff
@@ -190,11 +190,10 @@ public function getHungerManager() : HungerManager{
...
 			return false;
 		}
-
```
Comment (ShockedPlot7560) [inline src/entity/Human.php:196]: Unnecessary change

File: src/entity/Human.php (hunk 1 above)
Comment (ShockedPlot7560) [reply src/entity/Human.php]: Plus prefer conditions that consume less first like the variable check before the instance.

**Review by ShockedPlot7560 — CHANGES_REQUESTED** [review]

File: src/entity/Human.php (hunk 1 above)
Comment (dktapps) [reply src/entity/Human.php]: In practice the difference between the two is likely to be negligible, but in this case the instanceof is related to the later part of the condition also.

**Comment by dktapps** [comment]
Not a fan of the signature changes here. I think this should be doable without API changes.

File: src/entity/Human.php (hunk 1 above)
Comment (NTT1906) [reply src/entity/Human.php]: > Plus prefer conditions that consume less first like the variable check before the instance.

Can you explain it further?

**Comment by NTT1906** [comment]
By the way, how to automatically delete the use path when not using (e.g: use pocketmine\entity\Consumable;) in PhpStorm?

File: src/entity/Human.php [hunk 5]
```diff
@@ -191,8 +191,13 @@ public function getHungerManager() : HungerManager{
 	}
 
 	public function consumeObject(Consumable $consumable) : bool{
-		if($consumable instanceof FoodSource && $consumable->requiresHunger() && !$this->hungerManager->isHungry()){
-			return false;
+		if ($consumable instanceof FoodSource) {
+			if ($this instanceof Player && $this->allowForceEating()) {
```
Comment (dktapps) [inline src/entity/Human.php]: `$this instanceof` is a sure sign that your code design is flawed.

File: src/entity/Human.php [hunk 6]
```diff
@@ -170,6 +170,10 @@ public function sendSkin(?array $targets = null) : void{
 		]);
 	}
 
+	public function allowForceEating () : bool{
```
Comment (ShockedPlot7560) [inline src/entity/Human.php]:
```suggestion
	public function allowForceEating() : bool{
```

File: src/entity/Human.php [hunk 7]
```diff
@@ -191,8 +195,13 @@ public function getHungerManager() : HungerManager{
 	}
 
 	public function consumeObject(Consumable $consumable) : bool{
-		if($consumable instanceof FoodSource && $consumable->requiresHunger() && !$this->hungerManager->isHungry()){
-			return false;
+		if ($consumable instanceof FoodSource) {
```
Comment (ShockedPlot7560) [inline src/entity/Human.php]: Code style, no space before and after parentheses

File: src/entity/Human.php [hunk 8]
```diff
@@ -191,8 +195,13 @@ public function getHungerManager() : HungerManager{
...
+			if ($this->allowForceEating()) {
```
Comment (ShockedPlot7560) [inline src/entity/Human.php]: Same here

File: src/entity/Human.php [hunk 9]
```diff
@@ -191,8 +195,13 @@ public function getHungerManager() : HungerManager{
...
+				return parent::consumeObject($consumable);
+			}
+			if ($consumable->requiresHunger() && !$this->getHungerManager()->isHungry()) {
```
Comment (ShockedPlot7560) [inline src/entity/Human.php]: Same here

File: src/player/Player.php [hunk 10]
```diff
@@ -673,6 +673,10 @@ public function setUsingItem(bool $value) : void{
 		$this->networkPropertiesDirty = true;
 	}
 
+	public function allowForceEating () : bool{
```
Comment (ShockedPlot7560) [inline src/player/Player.php]:
```suggestion
	public function allowForceEating() : bool{
```

File: src/entity/Human.php [hunk 11]
```diff
@@ -191,8 +195,13 @@ public function getHungerManager() : HungerManager{
 	}
 
 	public function consumeObject(Consumable $consumable) : bool{
```
Comment (ShockedPlot7560) [inline src/entity/Human.php:196]: Can be simplified to have only two returns instead of duplicating the call to the parent

**Review by ShockedPlot7560 — CHANGES_REQUESTED** [review]
I find the name of the added function rather odd, but I wouldn't know how to replace it.

File: src/player/Player.php (hunk 10 above)
Comment (dktapps) [inline src/player/Player.php]: I'm not a fan of this name. It feels way too specific.

Comment (NTT1906) [reply src/player/Player.php]: > I'm not a fan of this name. It feels way too specific.

what is your idea?

Comment (NTT1906) [reply src/player/Player.php]: canPlayerConsume

canPlayerSnack
canPlayerChew
canPlayerDevour
isPlayerFamished

allowForceFeast
canFreelyEating

Comment (ShockedPlot7560) [reply src/player/Player.php]: Could `canEatAnytime` better ? It can't include `player` because it's linked to Human and not only Player.

Comment (dktapps) [reply src/player/Player.php]: None of these address my original comment.
I'm thinking more along the lines of `canEat()`, which covers the creative case and hunger together.

Comment (NTT1906) [reply src/player/Player.php]: `canEat()` seems too general, but I too go with that.

Comment (NTT1906) [reply src/player/Player.php]: @ShockedPlot7560 Can I go with "canEatWithoutHunger"?

Comment (NTT1906) [reply src/player/Player.php]: I think I will go with "canEatWithoutHunger"

**Review by dktapps — CHANGES_REQUESTED** [review]
I don't like this solution. I think a better solution would be to have a switch to enable/disable HungerManager, and a way for the World to notify players when the difficulty is changed. This would allow plugins more power over it rather than hardlocking the logic into a random `Player` function.

**Comment by dktapps** [comment]
Superseded by https://github.com/pmmp/PocketMine-MP/commit/12214792b32a63a0668a2ecbc175f148bd7fe222
